prov/hook: add a debug hooking provider

a-ilango opened the pull request.

shefty commented on prov/hook/perf/src/hook_perf.c:
```diff
@@ -883,22 +885,51 @@ static int perf_hook_fabric(struct fi_fabric_attr *attr,
+	ep->rma = &perf_rma_ops;
+	ep->tagged = &perf_tagged_ops;
+	return 0;
+}
```
The above calls can be static

shefty commented on prov/hook/perf/src/hook_perf.c:
```diff
@@ -883,22 +885,51 @@ static int perf_hook_fabric(struct fi_fabric_attr *attr,
 
 	hook_fabric_init(&fab->fabric_hook, HOOK_PERF, attr->fabric, hprov,
-			 &perf_fabric_fid_ops);
+			 &perf_fabric_fid_ops, &perf_prov_ctx);
```
I think we want to try replacing HOOK_PERF with a struct hook_provider * (now called struct hook_prov_ctx).

shefty commented on prov/hook/src/hook.c:
```diff
@@ -44,6 +44,53 @@ static char **hooks;
+		return (container_of(fid, struct hook_mr, mr.fid)->
+			domain->fabric);
+	default:
+		return NULL;
```
probably safe to assert(0) here

a-ilango replied:
This was the reason I was checking for NULL fabric in other places. If we force using debug hook only when --enable-debug=yes, then we can use assert(0) here.

shefty replied:
I don't understand how fabric can be null, regardless of which build/hook is used.  What situation causes this?  Do we still need to allocate 'hooked' objects for all underlying objects, even if we don't hook the calls for it?

shefty commented on prov/hook/src/hook_ep.c:
```diff
@@ -256,32 +246,57 @@ int hook_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
+	*ep = &myep->ep;
+	return 0;
+err:
+	fi_close(&myep->ep.fid);
```
We can't call fi_close() if fi_endpoint() fails.

shefty commented on prov/hook/src/hook_ep.c:
```diff
@@ -256,32 +246,57 @@ int hook_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
+	if (ret)
+		goto err;
+
+	ret = hook_ini_fid(dom->fabric->prov_ctx, &myep->ep.fid);
```
If a hook provider intercepts a create/close call, then any specified ini/fini call will not be invoked.  This is desired, but we should probably document that ini/fini calls are not invoked if the app hooks a create/close call.  The app must call ini/fini directly, or perform the desired work.

a-ilango replied:
There shouldn't be a need for ini/fini to be invoked if the hook provider intercepts create/close.

a-ilango commented on include/ofi_hook.h:
```diff
@@ -54,6 +55,7 @@
 	HOOK_PERF,
+	HOOK_DEBUG,
 	MAX_HOOKS
 };
```
I think this can be removed. I'll check.

a-ilango commented on include/ofi_hook.h:
```diff
@@ -66,19 +68,54 @@ enum ofi_hook_class {
+	struct fi_provider	prov;
+	int 			(*ini_fid[FI_CLASS_MAX])(struct fid *fid);
+	int 			(*fini_fid[FI_CLASS_MAX])(struct fid *fid);
+};
```
With just hook_provider, I couldn't come up with a good name for the field in the following struct. We already have a hprov. Do we just call it prov? That would break the convention of naming struct fi_provider field as prov.

```
struct hook_fabric {
        struct fid_fabric       fabric;
        struct fid_fabric       *hfabric;
        enum ofi_hook_class     hclass;
        struct fi_provider      *hprov;
        struct hook_prov_ctx    *prov_ctx;
};
```

shefty replied:
I think we could just call it prov.  That aligns with fabric/hfabric, etc.  For full symmetry, I guess you could store struct fi_provider instead of hook_provider, but I wouldn't actually do that.  We require your new struct.

Hmm... another option would be to store the ini/fini calls in a separate structure that we reference from struct fi_prov_context.  We could even extend the definition of fi_prov_context with a union that is accessed based on the prov_type.  That might work better if we want to support external hooks, without the external hook provider needing to implement everything.

a-ilango commented on prov/hook/src/hook_cq.c:
```diff
@@ -102,41 +102,54 @@ static struct fi_ops_cq hook_cq_ops = {
 
+	ret = hook_ini_fid(dom->fabric->prov_ctx, &mycq->cq.fid);
+	if (ret)
+		goto err2;
```
I suppose if the user provides ini, they should also provide fini. Is that what you require documenting?

a-ilango replied:
Btw, since on error we close only hcq not hook_cq->cq we may be good here.

shefty replied:
I was referring to if hook_ini_fid() fails, the error path calls hook_fini_fid().  That puts additional constraints on the provider to ensure that fini() can be called even if ini() does not complete successfully.

a-ilango replied:
The error path wouldn't call fini() as we're closing hcq directly. Only if we close mycq->cq, we would hit that.

shefty replied:
ah - got it

a-ilango commented:
Updated the PR addressing the comments. For some I've added TODOs. Also added a few more tracing. Tracking recvs doesn't work for now: need to debug a hang.

a-ilango commented:
@shefty this is ready for merge.

shefty commented on include/rdma/fabric.h:
```diff
@@ -485,6 +485,7 @@ enum {
 	FI_CLASS_CONNREQ,
 	FI_CLASS_MC,
 	FI_CLASS_NIC,
+	FI_CLASS_MAX,
```
This change needs to be dropped.  We don't want external apps trying to use this, since the intent is that the value would change between versions.

a-ilango replied:
Okay. The hook provider would need to define a FI_CLASS_MAX for ini and fini fid arrays. I'll make the changes.

a-ilango commented on include/ofi_hook.h:
```diff
@@ -67,18 +70,67 @@ extern struct fi_ops hook_fid_ops;
+	 * (e.g. fi_endpoint) directly, then these ini/fini calls won't be
+	 * invoked. */
+	int 			(*ini_fid[HOOK_FI_CLASS_MAX])(struct fid *fid);
+	int 			(*fini_fid[HOOK_FI_CLASS_MAX])(struct fid *fid);
```
Do these look okay?

shefty replied:
Yep -- I think we can do more work on this, but other changes can come later as needed.  My only concern was exposing the change in the public header file.  The atomic definitions had a 'LAST' enum value as well, and those ended up being an issue when adding the collective support.  So, now the 'LAST' values are no longer the last values...
